Add ProjectWideDependencyChecker `check` API (#206)
ember-cli/ember-cli-version-checker: README.md

    let checker = VersionChecker.forProject(this.project);

    for (let { name, root } = checker.allAddons()) {
      // access to the addon, in this case name and root
    }
  }
};
```

> ### check

Reviewer: ```suggestion

### check
```

> A utility to verify that addons are installed at appropriate versions.  `npm` and `yarn resolve conflicting transitive dependency requirements by installing multiple versions.  They do not include a mechanism for packages to declare that a dependency must be unique.  This is, however, a practical constraint when building Ember applications (for example, we would not want to build an application that shipped two versions of Ember Data).

Reviewer: ```suggestion
A utility to verify that addons are installed at appropriate versions.  `npm` and `yarn` resolve conflicting transitive dependency requirements by installing multiple versions.  They do not include a mechanism for packages to declare that a dependency must be unique.  This is, however, a practical constraint when building Ember applications (for example, we would not want to build an application that shipped two versions of Ember Data).
```

> A utility to verify that addons are installed at appropriate versions.  `npm` and `yarn resolve conflicting transitive dependency requirements by installing multiple versions.  They do not include a mechanism for packages to declare that a dependency must be unique.  This is, however, a practical constraint when building Ember applications (for example, we would not want to build an application that shipped two versions of Ember Data).

Reviewer: > They do not include a mechanism for packages to declare that a dependency must be unique.

FWIW, https://github.com/npm/rfcs/pull/23 might be a reasonable thing to link to that would actually make things nicer here.

    let checker = VersionChecker.forProject(this.project);

    for (let { name, root } = checker.allAddons()) {

>       // access to the addon, in this case root + name

Reviewer 2: Overly pedantic, I know, but just keeping the order consistent with the destructuring.

```suggestion
      // access to the addon, in this case name and root
```

    }
  }
};
```
### check

> A utility to verify that addons are installed at appropriate versions.
> Although `npm` and `yarn` typically manage dependencies for us, they do not
> provide a mechanism for an addon to specific a hard constraint. When `npm` or
> `yarn `see and invalid dependency, they resolve this by duplicating said
> dependency. In-general, the node ecosystem works fine with duplication;
> unfortunately, this strategy doesn't always work.

Reviewer 3: Suggestion that only fixes errors:

```suggestion
A utility to verify that addons are installed at appropriate versions.
Although `npm` and `yarn` typically manage dependencies for us, they do not
provide a mechanism for an addon to specific a hard constraint. When `npm` or
`yarn `see an invalid dependency, they resolve this by duplicating said
dependency. In-general, the node ecosystem works fine with duplication;
unfortunately, this strategy doesn't always work.
```

> A utility to verify that addons are installed at appropriate versions.
> Although `npm` and `yarn` typically manage dependencies for us, they do not
> provide a mechanism for an addon to specific a hard constraint. When `npm` or
> `yarn `see and invalid dependency, they resolve this by duplicating said
> dependency. In-general, the node ecosystem works fine with duplication;
> unfortunately, this strategy doesn't always work.

Reviewer 3: Alternative suggestion with more wordsmithing.

```suggestion
A utility to verify that addons are installed at appropriate versions.  `npm` and `yarn resolve conflicting transitive dependency requirements by installing multiple versions.  They do not include a mechanism for packages to declare that a dependency must be unique.  This is, however, a practical constraint when building Ember applications (for example, we would not want to build an application that shipped two versions of Ember Data).
```

> Every addon in the ember ecosystem implicitly depends `ember-source` and most

Reviewer 2: ```suggestion
Every addon in the ember ecosystem implicitly depends on `ember-source`, and most
```

> likely a specific version range. If that dependency is specified as a
> `package.json` dependency, a mismatch between application and addon would
> result in duplicating `ember-source`. This is both unhelpful and unwanted.

Reviewer 3: ```suggestion
likely a specific version range. If that dependency is specified as a
`package.json` dependency, a mismatch between application and addon would
result in duplicating `ember-source`.  Instead of failing the build, we would build an application with an unknown version of `ember-source`, subverting the point of specifying dependency version ranges in the first place!  The `check` API provides a mechanism to avoid this and fail fast in the build step, instead of building an invalid application with harder to debug runtime errors.
```

For example, as of today `ember-data` supports `ember-source` `>= 3.4.8`, if it
where to use this addon, it could specify this constraint and provide good
error messages to users.

> 
> ```
> const VersionChecker = require('ember-cli-version-checker');

Reviewer 3: ```suggestion

```javascript
const VersionChecker = require('ember-cli-version-checker');
```

module.exports = {
  name: 'awesome-addon',
  included() {
    this._super.included.apply(this, arguments);

    const checker = VersionChecker.forProject(this.project);
    const check = checker.check({
      'ember-source': '>= 3.4.8'
    });

    // if it would like to simply assert
    check.assert('[awesome-addon] dependency check failed');
    // will throw an error message similar to the following if the check was not satisfied:

    //   [awesome-addon] dependency check failed:
    //     - 'ember-source' expected version [>= 3.4.8] but got version: [2.0.0]

>     // if the check is more advanced, we can inspect the resulting check.

Reviewer 3: Avoid use of `check` in same sentence with two different meanings

```suggestion
    // if the requirements are more advanced, we can inspect the resulting check.
```
